### DISCUSSION TITLE
Server problems

### DISCUSSION THREAD

**Discussion by frankystone** [widelands/widelands#5203] [Q&A]
Actually the new dependency graphs for the documentation do not work correct and it seems this in an issue of the server. The server uses a systemd service-file to create the documentation once a day. But this service file does not create the svg-images, instead only texts are created. The source of a page shows the corresponding rst-directives are inside the rst-source.

Running the corresponding management command by hand works without any issues: The svg images are created.

Any ideas anyone?
I wrote stonerl a PM...

Since we have no server related issue tracker, i didn't know where to put this question/issue. We may need such an issue tracker somewhere. Any ideas where to put those questions?

**Comment by matthiakl** [comment]
It looks like the website documentation was generated with `make dirhtml` while we tested with `make html` (like in your provided zips).
Did you change this recently or are unaware of? Since this breaks the cross-references from within in the graphs, it might be related.

**Reply by frankystone** [reply]
Yes, i didn't thought of that. The result is that the links inside the svg images are wrong. I have a solution for that. Issue in on the way.

But this can't be the reason why it works nice when starting the command manually but not when using the systemd service file. Both use the same [management command](https://github.com/widelands/widelands-website/blob/35c2661f4fba521f802021258ea935aea6eb1883/documentation/management/commands/create_docs.py).

**Reply by matthiakl** [reply]
The only thing that comes to my mind is that the execution environment is different.
Therefore it might use a different python/sphinx version or the working directory is not correct (if relevant).

**Comment by frankystone** [comment]
Ah, found it:

Jan 13 02:53:51 widelands python[25097]: WARNING: dot command 'dot' cannot be run (needed for graphviz output), check the graphviz_dot setting

**Reply by frankystone** [reply]
> The only thing that comes to my mind is that the execution environment is different. Therefore it might use a different python/sphinx version or the working directory is not correct (if relevant).

Something like that. I have no clue about those service files. Here is the service file (exchanged the path):
~~~~
[Unit]
Description=Creating Source Code Documentation
OnFailure=notify-email@%i.service

[Service]
Type=oneshot
User=www-data
Group=www-data
Environment=VIRTUAL_ENV=/path/to/website/wlwebsite
Environment=PATH=/path/to/website/wlwebsite/bin
WorkingDirectory=/path/to/website/wlwebsite/code/widelands
ExecStart=/path/to/website/wlwebsite/bin/python manage.py create_docs
~~~~

**Reply by matthiakl** [reply]
Does `/path/to/website/wlwebsite/bin` contains all required tools (dot)? If not, you need to add the corresponding path to it
```
Environment=PATH="/path/to/website/wlwebsite/bin:/path/to/dot"
```

**Reply by frankystone** [reply]
we have /lib/python3.6/site-packages/pydot.py

Changed it like so:
~~~~
Environment=PATH=/path/to/website/wlwebsite/bin:/path/to/website/wlwebsite/lib/python3.6/site-packages
~~~~
and reloading units, start service -> no success

I thought always the virtual environment is somehow activated inside the service file...

**Reply by matthiakl** [reply]
This is only the python binding, but there must be a dot binary somewhere else. Run `which dot` to find it.

**Reply by frankystone** [reply]
Sorry for being slow ... 
~~~~
$:> which dot
/usr/bin/dot
~~~~
Looking at the path with activated virtualenv it is like that:
~~~~
$:> echo $PATH
/var/www/django_projects/wlwebsite/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
~~~~

Should i add this whole path to the service file? Just to be secure?

I am so scared when working on the server :-S

**Reply by matthiakl** [reply]
Yes, this looks like a normal PATH. You can try adding `/usr/bin` first, which should be sufficient to find dot. Or you add the full path. 
Alternatively you can call sphinx with the dot executable directly:
```
check_call(['sphinx-build',
                        '-b', builder,
                        '-d', os.path.join(self.sphinx_dir, 'build/doctrees'),
                        '-c', self.sphinx_conf_dir,
                        '-D', 'graphviz_dot=/usr/bin/dot',
                        os.path.join(self.sphinx_dir, 'source'),
                        self.build_dir,
                        ])
```
https://www.sphinx-doc.org/en/master/usage/extensions/graphviz.html#confval-graphviz_dot

**Reply by frankystone** [reply]
puhhhh....

Added it to the PATH as you suggested and: TADAAAAA -> works!

Many, many thanks! One problem solved :-)

You are also known as matthiakl on the website?

**Reply by matthiakl** [reply]
> You are also known as matthiakl on the website?

Yes.
I'm glad I could help you :-)
